### PR TITLE
 Add SleuthKafkaStreamsConfiguration to enable starting and reporting spans

### PR DESCRIPTION
**Pull request by timtebeek**
This PR enables starting and reporting new spans when handling Kafka Streams. It's a first draft, as I've not consulted beforehand if something like this would be merged if completed.

Things I was not sure about:
- is this functionality desired within spring-cloud-sleuth? I sure hope so!
- should this go inside or alongside `SleuthKafkaConfiguration`?
- should this support the transform/process/map functions in [KafkaStreamsTracing](https://github.com/apache/incubator-zipkin-brave/blob/master/instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java)?
- ~how to get log correlation to work with the threading model used?~
- what level of functionality and tests are required before a (first) merge can be considered?

I'd love to see Kafka Streams supported in full and out of the box. I hope this first attempt kick starts a discussion on how to achieve that, or if I should instead keep my developments separate from Spring Cloud Sleuth.

Update: This PR now provides an instance of [`KafkaStreamsTracing`](https://github.com/apache/incubator-zipkin-brave/blob/master/instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java) along with an `KafkaStreamsBuilderFactoryBeanPostProcessor` to allow for starting, activating and stopping new spans while processing Kafka Streams messages.

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+	@Override
+	public Object postProcessAfterInitialization(Object bean, String beanName)
+			throws BeansException {
+		if (bean instanceof StreamsBuilderFactoryBean) {
+			StreamsBuilderFactoryBean sbfb = (StreamsBuilderFactoryBean) bean;
+			if (log.isDebugEnabled()) {
+				log.debug("StreamsBuilderFactoryBean bean is auto-configured to enable tracing.");
+			}
+			sbfb.setClientSupplier(new SleuthKafkaClientSupplier(kafkaTracing));
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java]: Use `KafkaStreamsTracing.kafkaClientSupplier()` through https://github.com/apache/incubator-zipkin-brave/pull/887/files#diff-b8dd8e4445870a6a249cf383bdfd6cf7R54 once Brave 5.6.4 is used.

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = SleuthKafkaStreamsConfigurationTest.Config.class, webEnvironment = WebEnvironment.NONE)
+public class SleuthKafkaStreamsConfigurationTest {
+
+	@Autowired
+	TestTraceStreamsBuilderFactoryBean bean;
+
+	@Test
+	public void clientSupplierInvokedOnStreamsBuilderFactoryBean() {
+		Assert.assertTrue("StreamsBuilderFactoryBean#setClientSupplier(KafkaClientSupplier) not called",
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java]: Please use AssertJ, preferably via `BDDAssertions`

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import org.springframework.kafka.config.StreamsBuilderFactoryBean;
+
+/**
+ * {@link org.springframework.boot.autoconfigure.EnableAutoConfiguration
+ * Auto-configuration} enables Kafka Streams span creation and reporting.
+ *
+ * @author Tim te Beek
+ */
+@Configuration
+@ConditionalOnBean({ Tracing.class, KafkaStreamsConfiguration.class })
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java]: Why do we need `KafkaStreamsConfiguration` as a conditional on bean?

Comment (timtebeek) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java]: Hmm good catch, an earlier commit autowired a bean from there, but with the bean post processor approach I guess that's not needed anymore. I'll update accordingly.
```suggestion
@ConditionalOnBean(Tracing.class)
```

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java (hunk 2 above)
Comment (timtebeek) [reply spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java]: Will do first moment with access to a laptop, thanks!

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.cloud.sleuth.instrument.messaging.TraceMessagingAutoConfiguration.SleuthKafkaConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.config.KafkaStreamsConfiguration;
```
Comment (timtebeek) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfiguration.java]:
```suggestion
```

**Comment by adriancole** [comment]
fyi we have a brave 5.6.4 release (finally out) and there is a method which is relevant to this PR https://github.com/apache/incubator-zipkin-brave/releases/tag/5.6.4

I'll let @jeqo comment on the change itself as he made and uses the library in question :)

**Comment by timtebeek** [comment]
Waiting for Brave fixes in #1373

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = SleuthKafkaStreamsConfigurationTest.Config.class,
+		webEnvironment = WebEnvironment.NONE)
+public class SleuthKafkaStreamsConfigurationTest {
+
+	@Autowired
+	TestTraceStreamsBuilderFactoryBean streamsBuilderFactoryBean;
+
+	@Test
+	public void clientSupplierInvokedOnStreamsBuilderFactoryBean() {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java]: I would add javadoc about why it is important that StreamsBuilderFactoryBean has the tracing supplier. Maybe mention the use case that now works. cc @jeqo

Comment (timtebeek) [reply spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java]: Added documentation in src/main in d8338ba0cb0ef2dca37e11b2de6a8ef5b7925a79, hope that's clear enough to resolve this item.

Comment (adriancole) [reply spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/messaging/SleuthKafkaStreamsConfigurationTest.java]: yep

**Comment by adriancole** [comment]
thx

**Comment by timtebeek** [comment]
Thanks for the first round of feedback & fixing Brave so quickly; any input from others needed still before we can consider a merge?

The commit history has become a bit of a mess; I'd suggest a squash merge if possible, since the total change set is small. Or if desired I could start a fresh PR with a single commit containing just the current state.

Just let me know if there's anything I can do to bring this closer to a merge! :)

**Comment by adriancole** [comment]
I don't anticipate big feedback from this, and small things we can address post merge. Thanks very much for the work. Would be super cool to add an example of this in use as a branch here, like the others. basically switch the app to communicate with kafka using this https://github.com/openzipkin/sleuth-webmvc-example#going-further

**Comment by timtebeek** [comment]
Thanks a lot for helping out & merging; luckily all the hard stuff was done already in Brave. Looking forward to applying this at work!

I'll see if I can fit in example branches for both Spring Kafka and Kafka Streams, and update you there when I do.

**Comment by marcingrzejszczak** [comment]
Hey @timtebeek, thanks a lot for the great work and the PR!

**Pull request merged**